### DISCUSSION TITLE
Ccache not hitting for subsequent builds without modification in source code

### DISCUSSION THREAD

**Discussion by thiyagsdevops** [ccache/ccache#969] [Support]
We are planning to bring in ccache for our entire organization and during our POC stage we facing below issue,
**Issue**
Subsequent builds not hitting the cache even there is no change in the source code. But if we retry the same build then it hits the cache. Our goal is to reduce build timing using ccache.
Build A --> Fresh build
Build B --> Next build without source code change(NO CACHE hit)
Build B(Retry) --> Hits the cache
Build C --> Next build without source code change(NO CACHE hit)
Build C(Retry) --> Hits the cache
**BUILD SYSTEM**
Git lab
**CODE**
C++ and compiler clang
**TOOLS**
cmake with xcode
**SETUP**
set(CMAKE_XCODE_ATTRIBUTE_CC "$(SRCROOT)/tools/ccache-clang")
set(CMAKE_XCODE_ATTRIBUTE_CXX "$(SRCROOT)/tools/ccache-clang++")
ccache-clang--> exec ccache /usr/bin/clang "$@"
ccache-clang++--> exec ccache /usr/bin/clang++ "$@"
**BUILD A LOGS**
[2021-11-25T05:33:23.106053 31103] Trying direct lookup
[2021-11-25T05:33:23.106864 31103] No **1993gn0paokhorcfnc0pjsn6siki6aeta** in primary storage
[2021-11-25T05:33:23.108040 31103] Running preprocessor
[2021-11-25T05:33:23.108162 31103] Executing /usr/bin/clang -target x86_64-apple-macos10.14 -x c++ -E /dummypath/dummy.cpp

**BUILD B LOGS**(No change in source code)
[2021-11-25T05:33:23.106053 31103] Trying direct lookup
[2021-11-25T05:33:23.106864 31103] No **1cf8v42iga1p8t1t8s2888882qodgfqsi** in primary storage
[2021-11-25T05:33:23.108040 31103] Running preprocessor
[2021-11-25T05:33:23.108162 31103] Executing /usr/bin/clang -target x86_64-apple-macos10.14 -x c++ -E /dummypath/dummy.cpp

**BUILD B LOGS RETRY**(No change in source code)
[2021-11-25T06:40:12.004955 41506] Trying direct lookup
[2021-11-25T06:40:12.006263 41506] Retrieved **1cf8v42iga1p8t1t8s2888882qodgfqsi** from primary storage (/Users/dummy/Library/Caches/ccache/1/9/93gn0paokhorcfnc0pjsn6siki6aetaM)

**CCACHE CONFIG:**
debug = true
max_size = 100.0G
hard_link = true
sloppiness = include_file_mtime, include_file_ctime, time_macros, file_stat_matches
run_second_cpp = true

**Comment by jrosdahl** [comment]
Have you seen the [_Cache debugging_](https://ccache.dev/manual/4.5.1.html#_cache_debugging) part of the documentation? 

That is: to find out why BUILD A and BUILD B produces different hashes for `dummy.cpp` you should compare `dummy.o.ccache-input-text` from BUILD A with the same file from BUILD B.

**Reply by thiyagsdevops** [reply]
Thanks for the suggestion we found the root and by adding below in we fixed the issue. Now for each build the cache hits.
**export CCACHE_IGNOREOPTIONS="-DDUMMY_VERSION=*"**

**Comment by thiyagsdevops** [comment]
With below lines in CMakeLists.txt we were able to prefix ccache with clang compiler on mac with Xcode.
**set(CMAKE_XCODE_ATTRIBUTE_CC "$(SRCROOT)/tools/ccache-clang")
**set(CMAKE_XCODE_ATTRIBUTE_CXX "$(SRCROOT)/tools/ccache-clang++")**

ccache-clang++:
#!/bin/sh
if type -p ccache >/dev/null 2>&1; then
  #export CCACHE_MAXSIZE=100G
  export CCACHE_CPP2=true
  export CCACHE_HARDLINK=true
  export CCACHE_DEPEND=true
  export CCACHE_IGNOREOPTIONS="-DDUMMY_VERSION=*"
  export CCACHE_SLOPPINESS=file_macro,time_macros,include_file_mtime,include_file_ctime,file_stat_matches
  exec ccache clang++ "$@"
else
  exec clang++ "$@"
fi

Is there is a similar option available for windows with Visual Studio 2019 as well. Please suggest

**Reply by jrosdahl** [reply]
Please start a new discussion for this since it's a separate topic.
